File: iOS/DateMate/Data/DataSource/PostRemoteDataSource.swift
//
//  PostRemoteDataSource.swift
//  DateMate
//
//  Created by 홍희표 on 2025/11/21.
//

import Foundation
import FirebaseFirestore
import Combine

class PostRemoteDataSource {
    let firestore: Firestore
    
    func fetchUserPosts(userId: String) -> AnyPublisher<[Post], Never> {
        let mockPosts = [
            Post(id: "1", userId: userId,
                 title: "Traditional spare ribs baked",
                 content: "Chef John",
                 imageUrls: ["https://picsum.photos/400/200?random=1"], likeCount: 0),
            Post(id: "2", userId: userId,
                 title: "Spice roasted chicken with flavored rice",
                 content: "Mark Kelvin",
                 imageUrls: ["https://picsum.photos/400/200?random=2"], likeCount: 0),
            Post(id: "3", userId: userId,
                 title: "Spicy fried rice with bacon",
                 content: "Chef Anna",
                 imageUrls: ["https://picsum.photos/400/200?random=3"], likeCount: 0),
            Post(id: "4", userId: userId,
                 title: "Classic Beef Wellington",
                 content: "Gordon",
                 imageUrls: ["https://picsum.photos/400/200?random=4"], likeCount: 0)
            ]
        return Just(mockPosts)
            .delay(for: .milliseconds(500), scheduler: RunLoop.main)
            .eraseToAnyPublisher()
    }
    
    private init(_ firestore: Firestore) {
        self.firestore = firestore
    }
    
    private static var instance: PostRemoteDataSource? = nil
    
    static func getInstance(firestore: Firestore) -> PostRemoteDataSource {
        if let instance = self.instance {
            return instance
        } else {
            let postDataSource = PostRemoteDataSource(firestore)
            self.instance = postDataSource
            return postDataSource
        }
    }
}

File: Android/app/src/main/java/com/hhp227/datemate/data/datasource/PostRemoteDataSource.kt
package com.hhp227.datemate.data.datasource

import com.google.firebase.firestore.FirebaseFirestore
import com.hhp227.datemate.data.model.Post
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class PostRemoteDataSource private constructor(
    private val firestore: FirebaseFirestore
) {
    // 유저의 ID를 기반으로 게시물 목록을 가져옴 (현재는 Mock Data)
    fun fetchUserPosts(userId: String): Flow<List<Post>> = flow {
        delay(500) // 네트워크 지연 시뮬레이션

        val mockPosts = listOf(
            Post("1", userId, "Traditional spare ribs baked", "Chef John", listOf("https://picsum.photos/400/200?random=1")),
            Post("2", userId, "Spice roasted chicken with flavored rice", "Mark Kelvin", listOf("https://picsum.photos/400/200?random=2")),
            Post("3", userId, "Spicy fried rice with bacon", "Chef Anna", listOf("https://picsum.photos/400/200?random=3")),
            Post("4", userId, "Classic Beef Wellington", "Gordon", listOf("https://picsum.photos/400/200?random=4"))
        )
        emit(mockPosts)
    }

    companion object {
        @Volatile private var instance: PostRemoteDataSource? = null

        fun getInstance(firestore: FirebaseFirestore) =
            instance ?: synchronized(this) {
                instance ?: PostRemoteDataSource(firestore).also { instance = it }
            }
    }
}